src/librados/IoCtxImpl.cc

```diff
@@ -868,6 +868,57 @@ int librados::IoCtxImpl::aio_sparse_read(const object_t oid,
   return 0;
 }
 
+int librados::IoCtxImpl::aio_cmpext(const object_t& oid,
+				    AioCompletionImpl *c,
+				    uint64_t off,
+				    bufferlist& cmp_bl,
+				    bufferlist *mismatch_bl,
+				    uint64_t *mismatch_off, uint64_t snapid)
+{
+  if (cmp_bl.length() > UINT_MAX/2)
+    return -E2BIG;
+
+  Context *onack = new C_aio_Ack(c);
+
+  c->is_read = true;
+  c->io = this;
+
+  Objecter::Op *o = objecter->prepare_cmpext_op(
+    oid, oloc, off, cmp_bl, snapid, mismatch_bl, mismatch_off, 0,
+    onack, &c->objver);
+  objecter->op_submit(o, &c->tid);
+  return 0;
+}
+
+/* use m_ops.cmpext() + prepare_read_op() for non-bufferlist C API */
+int librados::IoCtxImpl::aio_cmpext(const object_t& oid,
+				    AioCompletionImpl *c,
+				    const char *cmp_buf, size_t cmp_len,
+				    uint64_t off,
+				    char *mismatch_buf, size_t *mismatch_len,
+				    uint64_t *mismatch_off, uint64_t snapid)
+{
+  if (cmp_len > UINT_MAX/2)
+    return -E2BIG;
+
+  bufferlist cmp_bl;
+  cmp_bl.append(cmp_buf, cmp_len);
+
+  Context *nested = new C_aio_Ack(c);
+  C_ObjectOperation *onack = new C_ObjectOperation(nested);
+
+  c->is_read = true;
+  c->io = this;
+
+  onack->m_ops.cmpext(off, cmp_len, cmp_buf, mismatch_buf, mismatch_len,
+		      mismatch_off, NULL);
+
+  Objecter::Op *o = objecter->prepare_read_op(
+    oid, oloc, onack->m_ops, snapid, NULL, 0, onack, &c->objver);
```

Comment:
FWIW, in case aio_sparse_read() is incorrectly ignoring the snapid parameter, I've pushed: https://github.com/ddiss/ceph/commits/librados_sparse_read_use_snapid
Will wait on feedback before submitting.